File: slides/number-theory/1_eratosthenes.cpp
#include <vector>
using namespace std;

vector<bool> eratosthenes(int n) {
    if (n < 2) n = 1;
    vector<bool> prime(n+1, true);
    prime[0] = prime[1] = false;
    for (int i = 2; i <= n; i++) {
        if (not prime[i]) continue;
        for (int j = i+i; j <= n; j += i) {
            prime[j] = false;
        }
    }
    return prime;
}


#include <iostream>
int main() {
    int n = 10;
    auto prime = eratosthenes(n);
    for (int i = 0; i <= n; i++) {
        cout << i << ' ' << prime[i] << endl;
    }
}

File: slides/number-theory/1_eratosthenes.py
def eratosthenes(n: int) -> list[bool]:
    if n < 2:
        n = 1
    prime = [True] * (n+1)
    prime[0] = prime[1] = False
    for i in range(2, n+1):
        if not prime[i]:
            continue
        for j in range(i+i, n + 1, i):
            prime[j] = False
    return prime


if __name__ == '__main__':
    n = 10
    prime = eratosthenes(n)
    for i in range(n+1):
        print(i, prime[i])
